redis/src/aio/connection.rs

```diff
@@ -502,3 +505,88 @@ pub(crate) async fn connect_simple<T: RedisRuntime>(
+/// a transaction loop until it succeeds. Similar to the synchronous [`transaction`](crate::transaction)
+/// function but for async execution.
+///
+/// The provided closure may be executed multiple times if the transaction fails due to
+/// watched keys being modified between WATCH and EXEC. Any side effects in the closure
+/// should account for possible multiple executions.
+///
+/// # Arguments
+///
+/// * `connection` - A connection implementing [`ConnectionLike`](crate::aio::ConnectionLike) + Clone
+/// * `keys` - Array of keys to watch for changes during the transaction
+/// * `func` - Closure that implements the transaction logic. Will receive a cloned connection
+///   and an atomic pipeline. Should return `Result<Option<T>, RedisError>` where:
+///   - `None` indicates the transaction should be retried
+///   - `Some(T)` contains the final result if successful
+///   - `Err` aborts the transaction
+///
+/// # Type Parameters
+///
+/// * `C` - Connection type implementing ConnectionLike + Clone
+/// * `K` - Key type implementing ToRedisArgs
+/// * `T` - Return value type implementing FromRedisValue
+/// * `F` - Closure type taking (C, Pipeline) and returning Fut
+/// * `Fut` - Future returning Result<Option<T>, RedisError>
+///
+/// # Examples
+///
+/// ```rust,no_run
+/// use redis::{AsyncCommands, RedisResult, pipe};
+///
+/// async fn increment(con: redis::aio::ConnectionManager) -> RedisResult<isize> {
+///     let key = "my_counter";
+///     redis::aio::transaction_async(con, &[key], |mut con, mut pipe| async move {
+///         //  Get the current value
+///         let val: isize = con.get(key).await?;
+///         //   Create pipeline that increments and returns new value
+///         pipe.set(key, val + 1)
+///             .ignore()
+///             .get(key)
+///             .query_async(&mut con)
+///             .await
+///     })
+///     .await
+/// }
+/// ```
+///
+/// # Notes
+///
+/// - The closure provided to this function may be executed multiple times if the
+///   watched keys are modified by other clients between the WATCH and EXEC commands
+/// - Side effects in the closure should be idempotent since they may occur multiple times
+/// - The transaction is automatically rolled back if the closure returns an error
+/// - Returns the final result once the transaction succeeds
+///
+/// For more details on Redis transactions, see the [Redis documentation](https://redis.io/topics/transactions)
+pub async fn transaction_async<
+    C: ConnectionLike + Clone,
+    K: ToRedisArgs,
+    T: FromRedisValue,
+    F: Fn(C, Pipeline) -> Fut,
+    Fut: Future<Output = Result<Option<T>, RedisError>>,
+>(
+    mut connection: C,
+    keys: &[K],
+    func: F,
+) -> Result<T, RedisError> {
+    loop {
+        cmd("WATCH").arg(keys).exec_async(&mut connection).await?;
+
+        let mut p = pipe();
+        let response = func(connection.clone(), p.atomic().to_owned()).await;
```

Comment:
instead of using an atomic pipeline, you should manually add before the closure the "MULTI" command, and after the closure the "EXEC" and "UNWATCH" commands.
That will save waiting for another back-and-forth between the client and the server for the UNWATCH.